receiver/awsecscontainermetricsreceiver/awsecscontainermetrics/label.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awsecscontainermetrics
+
+import (
+	"strings"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+)
+
+func containerLabelKeysAndValues(cm ContainerMetadata) ([]*metricspb.LabelKey, []*metricspb.LabelValue) {
```

Comment:
Do we need to add labels in the receiver? We usually add resource information in a processor, such as here

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/processor/k8sprocessor/kube/kube.go#L104

I think this would allow the same resource information to be added both to these metrics, and any other metrics / traces received by the collector.

Reply:
This is for ECS, not k8s.

Also, while Task labels could be added by a ECS Metadata processor, I am not sure container labels could be, since by the time you reach the processor you lose information about which container is which (unless you add labels that identify the container, in which case, you might as well just add all of them in the receiver directly).

Also, point #2- I am not sure the receiver can spit out container specific metrics without labels that identify which container they came from anyway. If you have 3 containers, and collect metrics for each, but don't differentiate them with something, then you are actually doing an aggregation across all the containers. At least that is my understanding.

I'll finally add that we've discussed this a bunch internally, and I think Rayhan has discussed it a bit externally as well. This implementation has been thought out.

Reply:
I feel like we should keep them as close as to the source (producer). Also, it will help to keep the config simple for customers. Customer will configure the `receiver` and good to go. 

I saw some other receivers did the same (like `kubelestatsreceiver`, `radisreceiver` and `statsdreceiver`).

Reply:
> This is for ECS, not k8s.

Yeah but k8s is an example with a presumably very similar pattern :)

> Also, while Task labels could be added by a ECS Metadata processor

From my understanding, it's not a could be but a when will it be - aren't we planning on adding an ECS processor so the information is added to data published by the application as well to allow correlation between ecs metrics and app metrics?

> container labels

Understood on the container labels - they do seem important to add in the receiver for those reasons. So yeah, the worst case is processor is also enabled and a small amount of extra overhead populating task information twice, but not a big deal.

> I'll finally add that we've discussed this a bunch internally, and I think Rayhan has discussed it a bit externally as well. This implementation has been thought out.

This isn't helpful because 

1) Internal discussions don't provide value when working in OSS. Going forward, it'll be good to move all related design discussions to GitHub or public Slack or gitter as well to faciliate AWS involvement in OpenTelemetry or other OSS
2) An implementation being thought out doesn't mean rubberstamped reviews or allow avoiding probing questions during reviews. It's what code reviews are there for

Reply:
> This isn't helpful because

Fair point 👍

> it's not a could be but a when will it be - aren't we planning on adding an ECS processor so the information is added to data published by the application as well to allow correlation between ecs metrics and app metrics?

I'm still confused by that actually. I heard the plan is that the OT SDKs will add the metadata themselves, and so there's not a huge need for the processor. I was told that doing it in the SDK is better- from within the container a call to Task Metadata can tell you which container you are a part of, so you can add the container specific labels. A collector processor would have trouble with that since once it reaches the collector you potentially lose the knowledge of which container in the task the application is.

I was told though that there might be some value in having a processor for other types of metrics/traces not from the OT SDKs. Like OpenCensus or Prometheus, since those won't have a built-in integration with ECS Metadata.

Reply:
>So yeah, the worst case is processor is also enabled and a small amount of extra overhead populating task information twice, but not a big deal.

I agree. Let's start with the minimum (putting them into the receiver for now). If we feel a strong need to add a processor, we can do that later.

Reply:
Yeah unfortunately the collector processor can't add docker container labels since the OTel collector is in a different container. But as long as its a sidecar it should be able to add the rest of the ECS labels. The current state is that the SDKs add container labels, which are standard since they're just docker - this can be provided by OTel in a generic way.

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/resource/semantic_conventions/container.md

They don't provide ECS-specific labels, mainly since they're not defined in the spec and aren't an SDK concern. We can add SDK extensions for ECS - and in the very long term, it makes sense to have SDK plugins for ECS labels too to support running the collector as a service instead of a sidecar. We see in practice that there is overlap between resource detection logic in the collector, like for GCP/EC2, and what's in the SDKs, primarily since collector as a sidecar provides an approach for underdeveloped or future languages while SDK implementations are more flexible in terms of deployment.

In the shorter term, if we restrict the conversation to sidecar as a first iteration, then implementing ECS labels as a processor would allow using from any app regardless of language, until ECS plugins could be added for each language too. For example, because Java and JS at least already populate the container info, if the collector has a processor for ECS tasks and above granularity, both languages would get the full set of ECS metadata automatically. So I think it makes sense to have a processor as the first implementation of ECS task metadata enrichment.

Hopefully that provides some overall context - but to avoid confusion, it's indeed unrelated to this PR as I realized that indeed it makes sense to additionally add them in the receiver :)

Reply:
I went through dockerstat receiver. They are adding labels to the metrics for each container: close to the source as mentioned by @hossain-rayhan 

1. ContainerStatsToMetrics: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/104a272746dfbda626c03a8f96e0885ebe559916/receiver/dockerstatsreceiver/metrics.go#L52
2.  **updateConfiguredResourceLabels**: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/104a272746dfbda626c03a8f96e0885ebe559916/receiver/dockerstatsreceiver/metrics.go#L80

Reply:
Right, these are associated with individual container stat requests so attaching them later would be generally impractical.  Host correlation or similar features not tied to metric generation would occur in a relevant processor though.